### PR TITLE
Sign request meta system

### PR DESCRIPTION
**Pull request by irfan798**

### CODE REVIEW

File: ur-packages/sign-request-response/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Signing Protocol
+
+This is the implementation of the [Signing Protocol](https://github.com/ngraveio/Research/blob/main/papers/nbcr-2023-003-crypto-sign.md) that is coin agnostic and can be used for any coin.
+
+This package add support for following ur types:
```
Comment (Pietahr) [inline ur-packages/sign-request-response/README.md]: "This package adds support for the following ur types:"

File: ur-packages/sign-request-response/README.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+| Type                      | [[CBOR Tag]](https://www.iana.org/assignments/cbor-tags/cbor-tags.xhtml) | Owner  | Description                                                                                               | Definition                                                                                                 |
+| ------------------------- | ------------------------------------------------------------------------ | ------ | --------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------- |
+| `crypto-sign-request` | 1411                                                                     | Ngrave | Import multiple accounts with and without output descriptors and specify optionally tokens to synchronize | [[NBCR-2023-003]](https://github.com/ngraveio/Research/blob/main/papers/nbcr-2023-003-crypto-sign.md) |
```
Comment (Pietahr) [inline ur-packages/sign-request-response/README.md]: "and specify optionally tokens to synchronize"
What do you mean with this explanation? Synchronise is no responsibility of this package?

File: ur-packages/sign-request-response/__test__/SignRequestTest.test.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
+import {
+  CryptoCoinIdentity,
+  CryptoKeypath,
+  EllipticCurve,
+  PathComponent,
+} from '@ngraveio/bc-ur-registry-crypto-coin-identity';
+import { CryptoTxSignature, CryptoSignRequest } from '../src';
+import { EthSignRequestMeta, IrfanSignRequestMeta, PolygonMeta } from '../src/metadatas/Ethereum.metadata';
```
Comment (Pietahr) [inline ur-packages/sign-request-response/__test__/SignRequestTest.test.ts]: IrfanSignRequestMeta 😄

File: ur-packages/sign-request-response/__test__/SignRequestTest.test.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      new PathComponent({ index: 0, hardened: false }),
+      new PathComponent({ index: 0, hardened: false }),
+    ]);
+
+    const ethSignRequest = new CryptoSignRequest({
+      requestId: Buffer.from('babecafe', 'hex'),
+      coinId: ethCoinId,
+      derivationPath: path,
+      signData: Buffer.from('abbacaca', 'hex'),
+      metadata: ethSignMeta, //polySignMeta, //ethSignMeta,
```
Comment (Pietahr) [inline ur-packages/sign-request-response/__test__/SignRequestTest.test.ts]: remove "//polySignMeta, //ethSignMeta"

File: ur-packages/sign-request-response/package.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "@ngraveio/bc-ur-sign-request-response",
```
Comment (Pietahr) [inline ur-packages/sign-request-response/package.json:2]: Both (response & request) will be part of the same package? to ensure compatibility?

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+      else if(requestId.length < 16) {
+        const padding = Buffer.alloc(16 - requestId.length);
+        this.requestId = Buffer.concat([requestId, padding]);
+      }
+      else this.requestId = requestId;
+    }
+    // If request id is not provided, generate a random one
+    else {
+      this.requestId = Buffer.from(require("crypto").randomBytes(16));
+    }
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts]: this logic can be moved to the setter of the requestId

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+    // Check inputs
+    CryptoSignRequest.checkInputs({
+      requestId,
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts:64]: the requestId input is only checked if it had the desired amount of characters, not when it is padded

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+          throw new Error(`Provided Metadata is an instance of ${currentType}, it should be instance of ${metaType.name} for coin ${coinId.toURL()}`);
+        }
+
+        this.metadata = metadata;
+      }
+      else {
+        // Create an general instance of foundmetadata type
+        this.metadata = new metaType(metadata);
+      }
+    }
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts]: this logic can be moved to the setter of the metadata

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+    derivationPath.setTag(this.derivationPath.getRegistryType().getTag());
+    map[Keys.derivationPath] = derivationPath;
+
+    map[Keys.signData] = this.signData;
+    if (this.masterFingerprint) map[Keys.masterFingerprint] = this.masterFingerprint;
+    if (this.origin) map[Keys.origin] = this.origin
+    if (this.metadata) map[Keys.metadata] = this.metadata.getData();
+
+    return new DataItem(map);;
+  };
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts]: docs

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+      if(signMetaMap.has(parentCoinId.toURL())) {
+        //console.debug("Found parent metadata for coin id", parentCoinId.toURL())
+        return signMetaMap.get(parentCoinId.toURL())!;
+      }
+    }
+
+    //console.debug('Fallback to general metadata')
+    // Otherwise return general metadata
+    return SignRequestMeta;
+  }
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts]: cleanup and docs

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+      coinId,
+      derivationPath,
+      signData,
+      masterFingerprint,
+      origin,
+      metadata
+    };
+
+    return new CryptoSignRequest(signRequestInput);
+  };
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoSignRequest.ts]: docs

File: ur-packages/sign-request-response/src/CryptoTxSignature.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    if(requestId) {
+      // Request id should not be longer than 16 bytes
+      if(requestId.length > 16) throw new Error("Request id should not be longer than 16 bytes");
+
+      // If request id is smaller than 16 bytes, pad with 0s
+      if(requestId.length < 16) {
+        const padding = Buffer.alloc(16 - requestId.length);
+        requestId = Buffer.concat([requestId, padding]);
+      }
+    }
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoTxSignature.ts]: this logic can be moved to the requestId setter

File: ur-packages/sign-request-response/src/CryptoTxSignature.ts [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+  };
+
+  public static fromDataItem = (dataItem: DataItem) => {
+    const map = dataItem.getData();
+    const signature = map[Keys.signature];
+    const requestId = map[Keys.requestId].getData();
+    const origin = map[Keys.origin];
+
+    return new CryptoTxSignature({signature, requestId, origin});
+  };
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/CryptoTxSignature.ts]: docs

File: ur-packages/sign-request-response/src/SignRequestMetadata.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+import { extend, DataItem, RegistryItem, DataItemMap } from '@keystonehq/bc-ur-registry'
+
+// Map metadata types by coinID
+// For coinID match, first check elliptic curve, then type, then subtype
+// If no match, return undefined
+
+// Also add types as interfaces for metadata
+
+
+// Add Patch functiomality
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/SignRequestMetadata.ts]: remove TODOs

File: ur-packages/sign-request-response/src/SignRequestMetadata.ts [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+
+export class SignRequestMeta extends DataItem {
+
+    coinId?: CryptoCoinIdentity = undefined;
+
+    constructor(data: DataItemMap, tag?: number) {
+        super(data, tag);
+        this.checkInputData(data);
+
+        //this.coinId = coinId;
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/SignRequestMetadata.ts]: remove

File: ur-packages/sign-request-response/src/SignRequestMetadata.ts [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    public toDataItem() {
+        return new DataItem(this.getData(), this.getTag());
+    }
+
+    // Throw error if data is undefined
+    public checkInputData(data: DataItemMap) {
+        if (typeof data !== 'object') {
+            throw new Error('Invalid input data');
+        }
+        // Extend this function to check for specific data types
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/SignRequestMetadata.ts]: still todo?

File: ur-packages/sign-request-response/src/metadatas/Ethereum.metadata.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    checkInputData(data: PolygonSignRequestProps) {
+        super.checkInputData(data);
+        // Make sure extraData is string
+        if (typeof data.extraData !== 'string') {
+            throw new Error('Invalid extraData');
+        }
+    }
+}
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/metadatas/Ethereum.metadata.ts]: move to a separate file

File: ur-packages/sign-request-response/src/metadatas/Ethereum.metadata.ts [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+
+export class IrfanSignRequestMeta extends SignRequestMeta {
+    static coinId: CryptoCoinIdentity = new CryptoCoinIdentity(EllipticCurve.Ed25519, 798);
+
+    constructor(data: object, tag?: number) {
+        super(data, tag);
+    }
+
+}
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/metadatas/Ethereum.metadata.ts]: delete ;)

File: ur-packages/sign-request-response/src/metadatas/index.ts [hunk 19]
```diff
@@ -0,0 +1,17 @@
+import { CryptoCoinIdentity, EllipticCurve } from "@ngraveio/bc-ur-registry-crypto-coin-identity";
+import { SignRequestMeta } from '../SignRequestMetadata';
+import { EthSignRequestMeta, PolygonMeta } from "./Ethereum.metadata";
+
+
+//export const signMetaMap = new Map<CryptoCoinIdentity, typeof SignRequestMeta>([
```
Comment (Pietahr) [inline ur-packages/sign-request-response/src/metadatas/index.ts:8]: remove

File: ur-packages/sign-request-response/README.md (hunk 2 above)
Comment (irfan798) [reply ur-packages/sign-request-response/README.md]: That description doesnt make sense at all, will change it

File: ur-packages/sign-request-response/__test__/SignRequestTest.test.ts (hunk 3 above)
Comment (irfan798) [reply ur-packages/sign-request-response/__test__/SignRequestTest.test.ts]: That is my current test type :D

File: ur-packages/sign-request-response/package.json (hunk 5 above)
Comment (irfan798) [reply ur-packages/sign-request-response/package.json:2]: yes

File: ur-packages/sign-request-response/src/CryptoSignRequest.ts (hunk 6 above)
Comment (irfan798) [reply ur-packages/sign-request-response/src/CryptoSignRequest.ts]: This seems correct to set `this.requestId` in constructor and give logic to set `_requestId` to setter. But typescript complains "Property '_requestId' has no initializer and is not definitely assigned in the constructor.ts(2564)"

File: ur-packages/sign-request-response/src/SignRequestMetadata.ts (hunk 16 above)
Comment (irfan798) [reply ur-packages/sign-request-response/src/SignRequestMetadata.ts]: This is a message for implementer of new SignRequestMeta classes, For example TRON needs to override and extend this functionlity

File: ur-packages/sign-request-response/src/metadatas/index.ts (hunk 19 above)
Comment (irfan798) [reply ur-packages/sign-request-response/src/metadatas/index.ts:8]: Actually i still want to it be like the commented line, but it didnt worked out as expected, i will try it again

**Pull request merged**